src/anm/animation/element.js

```diff
@@ -2570,45 +2570,50 @@ Element.getMatrixOf = function(elm, m) {
     return t;
 };
 
-Element.getTranslate = function(elm) {
-    if (elm.parent && elm.parent.layer2Bone) {
-	elm.$bonePath = elm.$bonePath || Element.bonePath(elm);
+Element.prototype.getTranslate = function() {
+    if (this.parent && this.parent.layer2Bone) {
+	var bonepath = this.bonePath();
         var result = {x: 0, y: 0};
+        var rotate = 0;
         var bone = null;
-	for (var li = elm.$bonePath.length; li--;) {
-            bone = elm.$bonePath[li];
-	    result.x += bone.bonelength * Math.cos(bone.bonerotate);
-	    result.y += bone.bonelength * Math.sin(bone.bonerotate);
+	for (var li = 0; li < bonepath.length; li++) {
+            bone = bonepath[li];
+            rotate += bone.bonerotate;
+	    result.x += bone.bonelength * Math.cos(rotate);
+	    result.y += bone.bonelength * Math.sin(rotate);
 	}
-        elm = bone ? elm.parent.children[bone.$from] : elm;
+        var elm = bone ? this.parent.children[bonepath[0].$from] : this;
         result.x += elm.x;
         result.y += elm.y;
         return result;
     }
-    return null;
+    return this;
 };
 
-Element.getRotate = function(elm) {
-    if (elm.parent && elm.parent.layer2Bone) {
-        elm.$bonePath = elm.$bonePath || Element.bonePath(elm);
+Element.prototype.getRotate = function() {
+    if (this.parent && this.parent.layer2Bone) {
+        this.$bonePath = this.$bonePath || Element.bonePath(this);
 	var result = 0;
-	for (var li = elm.$bonePath.length; li--;) {
-	    result += elm.$bonePath[li].bonerotate;
+	for (var li = this.$bonePath.length; li--;) {
+	    result += this.$bonePath[li].bonerotate;
 	}
-        return result;
+        return result + this.angle;
     }
-    return 0;
+    return this.angle;
 };
 
-Element.bonePath = function(elm) {
-    var l2b = elm.parent.layer2Bone;
-    var result = [];
-    var bone = l2b[elm.parent.children.indexOf(elm)];
-    while (bone) {
-       result.push(bone);
-       bone = l2b[bone.$from];
+Element.prototype.bonePath = function() {
+    if (!this.$bonePath) {
+        var l2b = this.parent.layer2Bone;
+        var result = [];
+        var bone = l2b[this.parent.children.indexOf(this)];
```

Comment:
Each element has an unique ID in `elm.id`, so you may use it to store in `layer2Bone`, instead of index, which actually could change in runtime in some rare cases.